tutorials/llm/llama/domain-adaptive-pretraining/code/tokenization_helper.py

```diff
@@ -0,0 +1,400 @@
+        dummy_tokens.append(f"<extra_id_{i}>")
+
+    record = []
+    N = len(m.pieces)
+    for i, sym in enumerate(new_tokens):
+        new_sym = m.SentencePiece()
+        new_sym.piece = sym
+        new_sym.score = 0.0  # default score for USER_DEFINED
+        new_sym.type = 4  # type value for USER_DEFINED
+        m.pieces.insert(N + i, new_sym)  # position after default control symbols ("<unk>", "<s>", "</s>")
+        record.append([sym, N + i])
+
+    N = len(m.pieces)
+    for i, sym in enumerate(dummy_tokens):
+        new_sym = m.SentencePiece()
+        new_sym.piece = sym
+        new_sym.score = 0.0  # default score for USER_DEFINED
+        new_sym.type = 4  # type value for USER_DEFINED
+        m.pieces.insert(N + i, new_sym)  # position after default control symbols ("<unk>", "<s>", "</s>")
+        record.append([sym, N + i])
+
+    print(f'New model pieces: {len(m.pieces)}')
+    print(m.trainer_spec)
+
+    with open(new_vocab_path, "w", encoding="utf8") as fp:
+        json.dump(record, fp)
+
+    with open(new_model_path, 'wb') as f:
+        f.write(m.SerializeToString())
+
+    if split > 1:
+        old_ebd_paths = []
+        for f in glob.glob(old_ebd_path + "/*.pt"):
+            old_ebd_paths.append(f)
+
+        def myFunc(s):
+            return int(s.split("embedding_")[-1].split(".")[0])  ### embedding_0.pt
+
+        old_ebd_paths.sort(key=myFunc)
+        word_embeddings = []
+        output_layers = []
+        for f in old_ebd_paths:
+            temp = torch.load(f)
+            word_embeddings.append(temp['word_embeddings'])
+            output_layers.append(temp['output_layer'])
+        word_embedding = torch.cat(word_embeddings, dim=1)
+        output_layer = torch.cat(output_layers, dim=0)
+        print("word_embedding shape: ", word_embedding.shape)
+        print("output_layer shape: ", output_layer.shape)
+
+        _, N = word_embedding.shape
+        add_weight = torch.zeros(add_cnt, N)
+        word_embedding = torch.cat((word_embedding, add_weight), 0)
+    else:
+        old_ebd = torch.load(old_ebd_path)
+        _, N = old_ebd['word_embeddings'].shape
+        add_weight = torch.zeros(add_cnt, N)
+        old_ebd['word_embeddings'] = torch.cat((old_ebd['word_embeddings'], add_weight), 0)
+
+    if split > 1:
+        _, M = output_layer.shape
+        add_out = torch.zeros(add_cnt, M)
+        output_layer = torch.cat((output_layer, add_out), 0)
+    else:
+        _, M = old_ebd['output_layer'].shape
+        add_out = torch.zeros(add_cnt, M)
+        old_ebd['output_layer'] = torch.cat((old_ebd['output_layer'], add_out), 0)
+
+    sp = spm.SentencePieceProcessor()
+    sp.load(original_tokenizer_path)
+
+    for r in record:
+        token = r[0]
+        idx = r[1]
+        ids = sp.encode_as_ids(token)
+        if split > 1:
+            word_embedding[idx] = torch.mean(word_embedding[ids], dim=0)
+            output_layer[idx] = torch.mean(output_layer[ids], dim=0)
```

Comment:
"The weights in the output layer corresponding to the new token are initialized to zero." according to the notebook?

Reply:
You are right. I have updated the notebook to reflect the mean initialization.